src/Features/Core/Portable/BraceCompletion/AbstractBraceCompletionService.cs

```diff
@@ -98,20 +112,21 @@ public virtual async Task<bool> CanProvideBraceCompletionAsync(char brace, int o
         }
 
         /// <summary>
-        /// Checks if the already inserted token is a valid opening token at the position in the document.
-        /// By default checks that the opening token is a valid token at the position and not in skipped token trivia.
+        /// Only called if <see cref="NeedsSemantics"/> returns true;
         /// </summary>
-        protected virtual Task<bool> IsValidOpenBraceTokenAtPositionAsync(SyntaxToken token, int position, Document document, CancellationToken cancellationToken)
+        protected virtual ValueTask<bool> IsValidOpenBraceTokenAtPositionAsync(Document document, SyntaxToken token, int position, CancellationToken cancellationToken)
         {
-            if (token.SpanStart != position)
-            {
-                return SpecializedTasks.False;
-            }
-
-            var syntaxFactsService = document.GetRequiredLanguageService<ISyntaxFactsService>();
-            return Task.FromResult(IsValidOpeningBraceToken(token) && !ParentIsSkippedTokensTriviaOrNull(syntaxFactsService, token));
+            // Subclass should have overridden this.
```

Comment:
why not making abstract?

Reply:
because 23 of the 24 subclasses would override and throw saying that this should not be called :)